### PR TITLE
feat: route-level entry generators

### LINKED ISSUE #9506: Specify which pages to prerender without relying on the crawler
### Describe the problem

By default, SvelteKit will find pages to prerender by starting at every parameter-less route (we call these _entries_), prerendering them (if possible), and crawling the resulting HTML to find new links.

You can specify new entries in `svelte.config.js` if you have pages that need to be prerendered but which are not, for whatever reason, findable by following `<a>` tags.

This works, but the ergonomics aren't super great, and there's no type safety.

### Describe the proposed solution

Next has a couple of solutions to this — [`getStaticPaths`](https://nextjs.org/docs/basic-features/data-fetching/get-static-paths) for old apps, and [`generateStaticParams`](https://beta.nextjs.org/docs/api-reference/generate-static-params) if you're using the App Router.

We could have our own version of `generateStaticParams` — for consistency with the configuration option, I'd propose `entries`:

```js
// src/routes/blog/[slug]/+page.server.js
export async function entries() {
  const posts = await get_posts_from_cms();

  return posts.map((post) => {
    return { slug: post.slug };
  });
}
```

The return value could be typed (including automatically, thanks to https://svelte.dev/blog/zero-config-type-safety).

In `entries` was specified for a page, the resulting entries would be included in `*`, meaning that they would be included by default, but you would still have total control over which entries were considered:

```js
export default {
  kit: {
    prerender: {
      // this is the default config, so can be omitted
      entries: ['*'] // include all parameter-less routes, plus pages that specify `entries`
    }
  }
};

export default {
  kit: {
    prerender: {
      entries: ['*', '/x/y/z'] // include all of the above _plus_ `/x/y/z`
    }
  }
};

export default {
  kit: {
    prerender: {
      entries: ['/x/y/z'] // only prerender `/x/y/z`
    }
  }
};
```

For consistency we should probably allow the same for `+server.js` files.

### PR DESCRIPTION
**Pull request by tcc-sejohnson**
Closes #9506 

This allows exporting an `entries` function from `+page`, `+page.server`, and `+server` files:

```ts
// /[date]/[...file]/+page.server.ts
export const entries = () => {
  return [
    { date: '2022-01-01', file: 'some/path.docx' },
    { date: '2022-02-01', file: 'some/other/path.docx' }
  ]
}
```

This would cause `/2022-01-01/some/path.docx` and `/2022-02-01/some/other/path.docx` to be prerendered, regardless of whether they're reachable by the prerendering crawler.

To aid in debugging and help make this feature as footgun-less as possible, if an `entries` export creates an entry that matches a route other than its own, it will throw an error during prerendering. I added a config option to handle this similarly to how we handle other prerender errors.

Given the following two routes:

- `/[slug]/[otherSlug]`
- `/[slug]/specific`

If `/[slug]/[otherSlug]/+page.server.ts` exports the following:

```ts
export const entries: EntryGenerator = () => {
  return [{ slug: 'foo', otherSlug: 'specific' }]
}
```

This would throw a prerendering error with a descriptive message, because the entry resolves to `/foo/specific`, which matches `[slug]/specific` instead of `/[slug]/[otherSlug]`. 

TODO:
- [x] Should we throw an error or something if routes generated from `entries` are duplicated? Should we just dedup them? 
- [x] Add typing support -- the return type should respect required and optional params
- [x] Make sure typing and functionality works for export from `+server.ts`
- [x] Tests
- [ ] Docs
- [x] Build auto-typing into the language server
- [ ] Fix the "You can't use this export" error my editor is still giving me???

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [x] It's really useful if your PR references an issue where it is discussed ahead of time. In many cases, features are absent for a reason. For large changes, please create an RFC: https://github.com/sveltejs/rfcs
- [x] This message body should clearly illustrate what problems it solves.
- [x] Ideally, include a test that fails without this PR but passes with it.

### Tests
- [x] Run the tests with `pnpm test` and lint the project with `pnpm lint` and `pnpm check`

### Changesets
- [x] If your PR makes a change that should be noted in one or more packages' changelogs, generate a changeset by running `pnpm changeset` and following the prompts. Changesets that add features should be `minor` and those that fix bugs should be `patch`. Please prefix changeset messages with `feat:`, `fix:`, or `chore:`.

### CODE REVIEW

File: packages/kit/src/utils/routing.js [hunk 1]
```diff
@@ -96,6 +96,61 @@ export function parse_route_id(id) {
 	return { pattern, params };
 }
 
+const basic_param_pattern = /\[(\[)?(?:\.\.\.)?(\w+?)(?:=(\w+))?\]\]?/;
+
+/**
+ * Parses a route ID, then resolves it to a route by replacing parameters with actual values from `entry`.
```
Comment (Rich-Harris) [inline packages/kit/src/utils/routing.js:102]:
```suggestion
 * Parses a route ID, then resolves it to a path by replacing parameters with actual values from `entry`.
```

File: packages/kit/src/utils/routing.js [hunk 2]
```diff
@@ -96,6 +96,61 @@ export function parse_route_id(id) {
...
+ * @param {string} id The route id
+ * @param {Record<string, string | undefined>} entry The entry meant to populate the route. For example, if the route is `/blog/[slug]`, the entry would be `{ slug: 'hello-world' }`
+ * @example
+ * ```js
+ * route_from_entry(`/blog/[slug]/[...somethingElse]`, { slug: 'hello-world', somethingElse: 'something/else' }); // `/blog/hello-world/something/else`
+ * ```
+ */
+export function route_from_entry(id, entry) {
```
Comment (Rich-Harris) [inline packages/kit/src/utils/routing.js:110]: the route is the input, not the output — we get a _path_ back from this function. maybe a name like this?

```suggestion
export function resolve_entry(id, entry) {
```

**Review by Rich-Harris — COMMENTED** [review]
this looks great, i just had one minor piece of feedback about naming. i think we probably need to ship the type updates first though, otherwise people will get yelled at if they try and use this

File: documentation/docs/20-core-concepts/40-page-options.md [hunk 3]
```diff
@@ -33,7 +33,7 @@ export const prerender = 'auto';
...
 
-The prerenderer will start at the root of your app and generate files for any prerenderable pages or `+server.js` routes it finds. Each page is scanned for `<a>` elements that point to other pages that are candidates for prerendering — because of this, you generally don't need to specify which pages should be accessed. If you _do_ need to specify which pages should be accessed by the prerenderer, you can do so with the `entries` option in the [prerender configuration](configuration#prerender).
+The prerenderer will start at the root of your app and generate files for any prerenderable pages or `+server.js` routes it finds. Each page is scanned for `<a>` elements that point to other pages that are candidates for prerendering — because of this, you generally don't need to specify which pages should be accessed. If you _do_ need to specify which pages should be accessed by the prerenderer, you can do so with the `entries` option in the [prerender configuration](configuration#prerender), or by exporting an entry generator from your dynamic route.
```
Comment (dummdidumm) [inline documentation/docs/20-core-concepts/40-page-options.md:36]: I think that's too little to add to the docs for this feature. We need some place to explain it properly.

Comment (tcc-sejohnson) [reply documentation/docs/20-core-concepts/40-page-options.md:36]: I agree -- I called this out in the PR description. I wasn't sure where we'd want to put this, though.

Comment (geoffrich) [reply documentation/docs/20-core-concepts/40-page-options.md:36]: Probably in [page options](https://kit.svelte.dev/docs/page-options) at the very least, and maybe also reference it in some of the prerender-related docs.

**Review by Rich-Harris — APPROVED** [review]

**Pull request merged**